go/plugins/googleai/googleai.go

```diff
@@ -16,17 +16,148 @@ package googleai
 
 import (
 	"context"
+	"errors"
 	"fmt"
+	"path"
+	"slices"
+	"sync"
 
 	"github.com/firebase/genkit/go/ai"
 	"github.com/firebase/genkit/go/plugins/internal/uri"
 	"github.com/google/generative-ai-go/genai"
+	"golang.org/x/exp/maps"
 	"google.golang.org/api/iterator"
 	"google.golang.org/api/option"
 )
 
-func newClient(ctx context.Context, apiKey string) (*genai.Client, error) {
-	return genai.NewClient(ctx, option.WithAPIKey(apiKey))
+var state struct {
+	mu        sync.Mutex
+	client    *genai.Client
+	models    map[string]ai.Generator
+	embedders map[string]ai.Embedder
+}
+
+// Config configures the plugin.
+type Config struct {
+	// API key. Required.
+	APIKey string
+	// Generative models to provide.
+	// If empty, a complete list will be obtained from the service.
+	Models []string
+	// Embedding models to provide.
+	// If empty, a complete list will be obtained from the service.
+	Embedders []string
+}
+
+func Init(ctx context.Context, cfg Config) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("googleai.Init: %w", err)
+		}
+	}()
+
+	if err := initClient(ctx, cfg.APIKey); err != nil {
+		return err
+	}
+
+	needModels := len(cfg.Models) == 0
+	needEmbedders := len(cfg.Embedders) == 0
+	if needModels || needEmbedders {
+		iter := state.client.ListModels(ctx)
+		for {
+			mi, err := iter.Next()
+			if err == iterator.Done {
+				break
+			}
+			if err != nil {
+				return err
+			}
+			// Model names are of the form "model/name".
+			name := path.Base(mi.Name)
+			if needModels && slices.Contains(mi.SupportedGenerationMethods, "generateContent") {
+				cfg.Models = append(cfg.Models, name)
+			}
+			if needEmbedders && slices.Contains(mi.SupportedGenerationMethods, "embedContent") {
+				cfg.Embedders = append(cfg.Embedders, name)
+			}
+		}
+	}
+	state.models = map[string]ai.Generator{}
+	for _, name := range cfg.Models {
+		state.models[name] = defineModel(name, state.client)
```

Comment:
I'm worried maintaining state in the plugin is passing on the burden to the plugin developer. As framework we really focus on ease of use for both plugins users and plugin developers. I think we should avoid state in the plugin.

When you call `ai.DefineGenerator` the defined generator gets added to the registry which already maintains the global state. `googleai.Generator` can just lookup the generator from registry directly... we could even make it `ai.Generator`, except the user would have to provide fully qualified model names `ai.Generate("googleai/gemini-1.5-flash")`

Reply:
Done.
And guess what, we do need LookupEmbedder, etc. after all!